### PR TITLE
Add implicit time marching to heat tutorial

### PR DESCRIPTION
**Pull request by charleskawczynski**
# Description

This PR makes the heat tutorial use an implicit time marching method (`Kvaerno3`) from the DiffEq library and adds an implicit solver type to the driver configuration.

At the moment, we cannot use ForwardDiff to approximate the Jacobians. Instead, DiffEq is falling back on finite difference method, which results in more kernel calls. Next, we'd like to see how this implicit method fairs with explicit for a range of thermal conductivities (which varies the stiffness). While I'm working on a preliminary comparison PR, @kanav99 will be trying to get ForwardDiff working with #1291 to reduce the number of kernel calls. Thanks @kanav99 and @thomasgibson for all of your help with time marching!

I have

- [ ] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [ ] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [ ] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [ ] There are no open pull requests for this already
- [ ] CLIMA developers with relevant expertise have been assigned to review this submission
- [ ] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

**Comment by jkozdon** [comment]
I mentioned this to @charleskawczynski on slack, but I believe that `ROCK4` is explicit, albeit designed for diffusion (i.e., stability region includes a lot of real axis)

File: src/Driver/SolverTypes/ImplicitSolverType.jl [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+"""
+    getdtmodel(ode_solver::AbstractSolverType, bl)
+
+A function which returns a model representing the dynamics
+with the most restrictive time-stepping requirements.
+"""
+function getdtmodel(ode_solver::ImplicitSolverType, bl)
+    # For implicit methods, the entire model itself
+    # contributes to the total stability of the time-integrator
+    return getdt(ode_solver.solver_method)
```
Comment (thomasgibson) [inline src/Driver/SolverTypes/ImplicitSolverType.jl:40]: This should return a balance law. This is ultimately used in the `calculate_dt` function in the main driver.

File: src/Driver/SolverTypes/ImplicitSolverType.jl [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+function solversetup(
+    ode_solver::ImplicitSolverType,
+    dg,
+    Q,
+    dt,
+    t0,
+    diffusion_direction,
+)
+
+    solver = ode_solver.solver_method(dg, Q; dt = dt, t0 = t0)
```
Comment (thomasgibson) [inline src/Driver/SolverTypes/ImplicitSolverType.jl:66]: How do I tell this solver that I want to use a different nonlinear/linear implicit method?

File: src/Driver/SolverTypes/ImplicitSolverType.jl [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    solver.
+    Default: `KenCarp4`
+"""
+struct ImplicitSolverType <: AbstractSolverType
+    # Function for an implicit method
+    solver_method::Function
+
+    function ImplicitSolverType(
+        alg;
+        solver_method = ODESolvers.DiffEqJLConstructor(alg),
```
Comment (thomasgibson) [inline src/Driver/SolverTypes/ImplicitSolverType.jl:24]: I think this answers my initial question: `alg` contains the timestepping info, and I'm assuming this also somehow attaches information about the actual implicit solver algorithm?

File: src/Driver/SolverTypes/ImplicitSolverType.jl (hunk 2 above)
Comment (charleskawczynski) [reply src/Driver/SolverTypes/ImplicitSolverType.jl:66]: ⬇️

**Comment by charleskawczynski** [comment]
Okay, updated to the implicit `Kvaerno3` method.

File: src/Arrays/MPIStateArrays.jl [hunk 4]
```diff
@@ -748,6 +748,8 @@ array_device(::Union{Array, SArray, MArray}) = CPU()
 array_device(::CuArray) = CUDADevice()
 array_device(s::SubArray) = array_device(parent(s))
 array_device(Q::MPIStateArray) = array_device(Q.data)
+array_device(ra::Base.ReshapedArray{T, N, A}) where {T, N, A <: MPIStateArray} =
+    array_device(ra.parent)
```
Comment (kpamnany) [inline src/Arrays/MPIStateArrays.jl:752]:
```suggestion
    array_device(parent(ra))
```
Maybe? Not sure. `Base.ReshapedArray` is a `UnionAll`...

Comment (charleskawczynski) [reply src/Arrays/MPIStateArrays.jl:752]: That should work fine. I see `parent(A::ReshapedArray) = A.parent` in `julia/base/reshapedarray.jl`

**Pull request merged**